### DISCUSSION TITLE
How to animate a scene?

### DISCUSSION THREAD

**Discussion by clairexen** [jimy-byerley/pymadcad#39] [Q&A] [answered]
Is there a simple way to set something like a timer callback that's being called from `show()` that allows me to update the scene in regular intervals?

**Comment by jimy-byerley** [comment, marked as answer]
Hello @clairexen !
`show()` is not the way to do this. It is just a helper to make a view window on a scene and run Qt if not already started.

To make fancy things, like animations, events, callbacks, you need to get to the Qt level of things
Taking a quick look at [`show`](https://github.com/jimy-byerley/pymadcad/blob/01d4eead22d2a695162d2baf2f98bd19887e622f/madcad/rendering.py#L64) you can see that a madcad 3d view is a regular Qt widget, so you can build a window out of it, and use [`QTimer`](https://doc.qt.io/qt-5/qtimer.html) time update it periodically

### To update the view periodically

You can schedule a rendering at any moment using `View.update()` (like any QWidget)
Depending on what you are looking for, you can trigger it periodically, or when you have changed something in the scene

```python
app = QApplication()

# build the scene itself (can be shared through views, also it is GUI agnostic
scene = madcad.rendering.Scene([...objs...], options={...})  # put here what you would usually put in `show()` 

# build the view widget
view = madcad.rendering.View(scene)
view.show()
# build the timer to refresh the view
timer = QTimer()
timer.setInterval(100)
timer.timeout.connect(view.update)
timer.start()

# run the event loop
app.exec()
```

### To animate things

`madcad.rendering` does not provide any specific tool to animate. However you can change the attributes  of some [displays](https://github.com/jimy-byerley/pymadcad/blob/master/madcad/displays.py) every frame and refresh the view. For instance, having a `PointDisplay` you can write the point color and position every frame.

If you are familiar with openGL, you can also write your own displays and use `time.time()` to compute the animation state every frame.

I hope this helps !
